pkg/sql/colexec/shuffleV2/shuffle.go

```diff
@@ -0,0 +1,729 @@
+		result.Batch = shuffle.ctr.shufflePool.getEndingBatch(shuffle.ctr.buf, shuffle.CurrentShuffleIdx, proc, shuffle.IsDebug)
+		shuffle.ctr.buf = result.Batch
+		return result, nil
+	}
+
+	var err error
+	for {
+		tmpBat := shuffle.ctr.shufflePool.getFullBatch(shuffle.ctr.buf, shuffle.CurrentShuffleIdx)
+		if tmpBat != nil && tmpBat.RowCount() > 0 { // find a full batch
+			shuffle.ctr.buf = tmpBat
+			break
+		}
+		// do input
+		result, err = vm.ChildrenCall(shuffle.GetChildren(0), proc, analyzer)
+		if err != nil {
+			return result, err
+		}
+		bat := result.Batch
+		if bat == nil {
+			shuffle.ctr.ending = true
+			shuffle.ctr.shufflePool.stopWriting()
+			goto SENDLAST
+		} else if !bat.IsEmpty() {
+			if shuffle.ShuffleType == int32(plan.ShuffleType_Hash) {
+				bat, err = hashShuffle(shuffle, bat, proc)
+			} else if shuffle.ShuffleType == int32(plan.ShuffleType_Range) {
+				bat, err = rangeShuffle(shuffle, bat, proc)
+			}
+			if err != nil {
+				return result, err
+			}
+			if bat != nil {
+				// can directly send this batch
+				shuffle.ctr.shufflePool.statsDirectlySentBatch(bat)
+				return result, nil
+			}
+		}
+	}
+	// send the batch
+	result.Batch = shuffle.ctr.buf
+	return result, nil
+}
+
+func (shuffle *ShuffleV2) clearSels() [][]int32 {
+	for i := range shuffle.ctr.sels {
+		if len(shuffle.ctr.sels[i]) > 0 {
+			shuffle.ctr.sels[i] = shuffle.ctr.sels[i][:0]
+		}
+	}
+	return shuffle.ctr.sels
+}
+
+func shuffleConstVectorByHash(ap *ShuffleV2, bat *batch.Batch) uint64 {
+	lenRegs := uint64(ap.BucketNum)
+	groupByVec := bat.Vecs[ap.ShuffleColIdx]
+	switch groupByVec.GetType().Oid {
+	case types.T_int64:
+		groupByCol := vector.MustFixedColNoTypeCheck[int64](groupByVec)
+		return plan2.SimpleInt64HashToRange(uint64(groupByCol[0]), lenRegs)
+	case types.T_int32:
+		groupByCol := vector.MustFixedColNoTypeCheck[int32](groupByVec)
+		return plan2.SimpleInt64HashToRange(uint64(groupByCol[0]), lenRegs)
+	case types.T_int16:
+		groupByCol := vector.MustFixedColNoTypeCheck[int16](groupByVec)
+		return plan2.SimpleInt64HashToRange(uint64(groupByCol[0]), lenRegs)
+	case types.T_uint64:
+		groupByCol := vector.MustFixedColNoTypeCheck[uint64](groupByVec)
+		return plan2.SimpleInt64HashToRange(uint64(groupByCol[0]), lenRegs)
+	case types.T_uint32:
+		groupByCol := vector.MustFixedColNoTypeCheck[uint32](groupByVec)
+		return plan2.SimpleInt64HashToRange(uint64(groupByCol[0]), lenRegs)
+	case types.T_uint16:
+		groupByCol := vector.MustFixedColNoTypeCheck[uint16](groupByVec)
+		return plan2.SimpleInt64HashToRange(uint64(groupByCol[0]), lenRegs)
+	case types.T_char, types.T_varchar, types.T_text:
+		groupByCol, area := vector.MustVarlenaRawData(groupByVec)
+		return plan2.SimpleCharHashToRange(groupByCol[0].GetByteSlice(area), lenRegs)
+	default:
+		panic("unsupported shuffle type, wrong plan!") //something got wrong here!
+	}
+}
+
+func getShuffledSelsByHashWithNull(ap *ShuffleV2, bat *batch.Batch) [][]int32 {
```

Comment:
What is the difference between WithNull and WithoutNull?

And it is much better to have a hashable generics based on types.T

Reply:
So should be later by range, I think what you need is a comparable generics.